### PR TITLE
Priv val via sockets

### PR DESCRIPTION
**Pull request by ebuchman**
Support a socket client and server for the priv validator.

Dead simple synchronous server.

@xla @ethanfrey want to work on this ? 

see https://github.com/tendermint/tendermint/pull/1044 for more context

### CODE REVIEW

**Comment by xla** [comment]
@ebuchman Can take this on after wrapping up #1056

**Comment by adrianbrink** [comment]
implements #819

File: types/priv_validator/socket.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+func readWrite(conn net.Conn, req PrivValidatorSocketMsg) (res PrivValidatorSocketMsg, err error) {
+	b := wire.BinaryBytes(req)
+	_, err = conn.Write(b)
+	if err != nil {
+		return nil, err
+	}
+
+	var n int
+	b = wire.ReadByteSlice(conn, 0, &n, &err) //XXX: no max
```
Comment (melekes) [inline types/priv_validator/socket.go]: check for err after before calling `decodeMsg`

File: types/priv_validator/socket.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+	return decodeMsg(b)
+}
+
+func decodeMsg(bz []byte) (msg PrivValidatorSocketMsg, err error) {
+	n := new(int)
+	r := bytes.NewReader(bz)
+	msgI := wire.ReadBinary(struct{ PrivValidatorSocketMsg }{}, r, 0, n, &err)
+	msg = msgI.(struct{ PrivValidatorSocketMsg }).PrivValidatorSocketMsg
```
Comment (melekes) [inline types/priv_validator/socket.go]: is msgI guaranteed to be PrivValidatorSocketMsg in case of error?

File: types/priv_validator/socket.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+				return // Ignore error from listener closing.
+			}
+			pvss.Logger.Error("Failed to accept connection: " + err.Error())
+			continue
+		}
+
+		pvss.Logger.Info("Accepted a new connection")
+
+		// read/write
+		for {
```
Comment (melekes) [inline types/priv_validator/socket.go]: only one connection at a time?

File: types/priv_validator/socket.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+		addr:    addr,
+		privVal: privVal,
+		chainID: chainID,
+	}
+	pvss.BaseService = *cmn.NewBaseService(logger, "privValidatorSocketServer", pvss)
+	return pvss
+}
+
+func (pvss *PrivValidatorSocketServer) OnStart() error {
+	if err := pvss.BaseService.OnStart(); err != nil {
```
Comment (melekes) [inline types/priv_validator/socket.go]: BaseService#OnStart is empty. no need to call it.

File: types/priv_validator/socket.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+	if err != nil {
+		return err
+	}
+	pvss.listener = ln
+	go pvss.acceptConnectionsRoutine()
+	return nil
+}
+
+func (pvss *PrivValidatorSocketServer) OnStop() {
+	pvss.BaseService.OnStop()
```
Comment (melekes) [inline types/priv_validator/socket.go]: BaseService#OnStop is empty. no need to call it.

File: types/priv_validator/socket.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+	if err := pvss.listener.Close(); err != nil {
+		pvss.Logger.Error("Error closing listener", "err", err)
+	}
+
+	if err := pvss.conn.Close(); err != nil {
```
Comment (melekes) [inline types/priv_validator/socket.go]: check if `pvss.conn != nil`

File: types/priv_validator/socket.go (hunk 3 above)
Comment (xla) [reply types/priv_validator/socket.go]: @ebuchman Is one connection at a time by design, or should the server be able to hanlde multiple, if so are there any upper bounds we want to enforce, or any other limiting requirements?

**Comment by ebuchman** [comment]
Thanks Alex.

Couple notes here. 

- Server should be on the Tendermint side. The signer should dial Tendermint.
- Should be able to handle multiple connections, but let's cap it at something reasonable (maybe 3). 
- Let's use go-wire for serialization
- Let's use the p2p.SecretConnection for authentication

@tarcieri is working on the signer itself in Rust

**Comment by ebuchman** [comment]
> Should the server exist as it's own binary under cmd/priv_validator or ultimately become a command of the tendermint cli?

No it should run in-process when `tendermint node` boots up. A tendermint node always needs a priv validator - either it's given a priv_validator.json with a private key direcetly (ie. useful for dev/testing) or it starts an in-process listener and waits to receive a connection from an external signing process

> Going to sync with him to see if we can make each others work easier.

Sounds great

**Comment by xla** [comment]
@ebuchman @melekes This is ready for review from my perspective. Tried to address all points.

When working with the protocol between client and server I was dearly missing some form of communicating errors,  but I don't know how settle we are on the current protocol.

**Comment by ebuchman** [comment]
Thanks Alexis. I've moved this over to a [fresh PR](https://github.com/tendermint/tendermint/pull/1204) by rebasing on a fresh copy of the new validator code without integrating to the rest of the codebase. Let's take up review/discussion there

**Pull request closed without merging**